src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -13432,7 +13432,12 @@ bool shouldParseSeparatorOrElement()
                 // We're not on a valid separator, but we want to be resilient for the user accidentally using the wrong
                 // one in common cases.
                 if (allowSemicolonAsSeparator && this.CurrentToken.Kind is SyntaxKind.SemicolonToken)
-                    return true;
+                {
+                    using var semicolonResetPoint = GetDisposableResetPoint(resetOnDispose: true);
+                    this.EatToken();
+                    bool canContinue = this.CurrentToken.Kind == closeTokenKind || isPossibleElement(this);
+                    return canContinue;
+                }
```

Comment:
sorry, i don't know how to make it clearer.  you say this code is needed, otherwise some other cases get worse.  But i'm saying: fix up the normal object-init parsing code to not make those cases worse.  Also, do we have tests for those cases somewhere?  If so, can you tell me their names?